auth/oidc.go

```diff
@@ -212,73 +238,148 @@ func (op *OIDCProvider) DiscoverConfig() (config *oidc.ProviderConfig, shouldSyn
 	return config, shouldSync, err
 }
 
-func (op *OIDCProvider) FetchCustomProviderConfig(discoveryURL string) (*oidc.ProviderConfig, error) {
-
-	var customConfig OidcProviderConfiguration
+func GetOIDCUsername(provider *OIDCProvider, subject string) string {
+	return fmt.Sprintf("%s_%s", provider.UserPrefix, url.QueryEscape(subject))
+}
 
+// Clone of provider metadata construction from go-oidc library.
+// https://github.com/coreos/go-oidc/blob/v2/oidc.go#L113
+func (op *OIDCProvider) FetchCustomProviderConfig(ctx context.Context, discoveryURL string) (*oidc.Provider, error) {
 	// If discovery URL is empty, use the standard discovery URL
 	if discoveryURL == "" {
 		discoveryURL = strings.TrimSuffix(op.Issuer, "/") + discoveryConfigPath
 	}
 
 	base.Debugf(base.KeyAuth, "Fetching custom provider config from %s", base.UD(discoveryURL))
-	req, err := http.NewRequest("GET", discoveryURL, nil)
+	req, err := http.NewRequest(http.MethodGet, discoveryURL, nil)
 	if err != nil {
 		base.Debugf(base.KeyAuth, "Error building new request for URL %s: %v", base.UD(discoveryURL), err)
 		return nil, err
 	}
-	resp, err := http.DefaultClient.Do(req)
+	resp, err := doRequest(ctx, req)
 	if err != nil {
 		base.Debugf(base.KeyAuth, "Error invoking calling discovery URL %s: %v", base.UD(discoveryURL), err)
 		return nil, err
 	}
-	defer func() { _ = resp.Body.Close() }()
-	if err := base.JSONDecoder(resp.Body).Decode(&customConfig); err != nil {
-		base.Debugf(base.KeyAuth, "Error parsing body %s: %v", base.UD(discoveryURL), err)
-		return nil, err
+	defer func() {
+		if err := resp.Body.Close(); err != nil {
+			base.Debugf(base.KeyAuth, "Error closing response body %s: %v", base.UD(discoveryURL), err)
+		}
+	}()
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		base.Debugf(base.KeyAuth, "Error reading response body %s: %v", base.UD(discoveryURL), err)
+		return nil, fmt.Errorf("unable to read response body: %v", err)
 	}
 
-	var oidcConfig oidc.ProviderConfig
-	oidcConfig, err = customConfig.AsProviderConfig()
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("%s: %s", resp.Status, body)
+	}
+
+	var p ProviderMetadata
+	err = unmarshalResp(resp, body, &p)
```

Comment:
Was it necessary to switch from the previous Decode approach?  Adding the status check seems like a reasonable enhancement, but I'm less sure about the other changes in unmarshalResp.  It looks like it's just providing some additional Content-Type error information which doesn't really provide particular value.